src/main/scala/beam/router/skim/ActivitySimSkimmer.scala

```diff
@@ -0,0 +1,273 @@
+
+  override protected def aggregateWithinIteration(
+    prevObservation: Option[AbstractSkimmerInternal],
+    currObservation: AbstractSkimmerInternal
+  ): AbstractSkimmerInternal = {
+    val prevSkim = prevObservation
+      .map(_.asInstanceOf[ActivitySimSkimmerInternal])
+      .getOrElse(ActivitySimSkimmerInternal.empty)
+    val currSkim = currObservation.asInstanceOf[ActivitySimSkimmerInternal]
+
+    def aggregatedDoubleSkimValue(getValue: ActivitySimSkimmerInternal => Double): Double = {
+      (getValue(prevSkim) * prevSkim.observations + getValue(currSkim) * currSkim.observations) / (prevSkim.observations + currSkim.observations)
+    }
+
+    ActivitySimSkimmerInternal(
+      travelTimeInMinutes = aggregatedDoubleSkimValue(_.travelTimeInMinutes),
+      generalizedTimeInMinutes = aggregatedDoubleSkimValue(_.generalizedTimeInMinutes),
+      generalizedCost = aggregatedDoubleSkimValue(_.generalizedCost),
+      distanceInMeters = aggregatedDoubleSkimValue(_.distanceInMeters),
+      cost = aggregatedDoubleSkimValue(_.cost),
+      energy = aggregatedDoubleSkimValue(_.energy),
+      walkAccessInMinutes = aggregatedDoubleSkimValue(_.walkAccessInMinutes),
+      walkEgressInMinutes = aggregatedDoubleSkimValue(_.walkEgressInMinutes),
+      walkAuxiliaryInMinutes = aggregatedDoubleSkimValue(_.walkAuxiliaryInMinutes),
+      totalInVehicleTimeInMinutes = aggregatedDoubleSkimValue(_.totalInVehicleTimeInMinutes),
+      driveTimeInMinutes = aggregatedDoubleSkimValue(_.driveTimeInMinutes),
+      driveDistanceInMeters = aggregatedDoubleSkimValue(_.driveDistanceInMeters),
+      ferryInVehicleTimeInMinutes = aggregatedDoubleSkimValue(_.ferryInVehicleTimeInMinutes),
+      lightRailInVehicleTimeInMinutes = aggregatedDoubleSkimValue(_.lightRailInVehicleTimeInMinutes),
+      transitBoardingsCount = aggregatedDoubleSkimValue(_.transitBoardingsCount),
+      observations = prevSkim.observations + currSkim.observations,
+      iterations = matsimServices.getIterationNumber + 1,
+      debugText = Seq(prevSkim.debugText, currSkim.debugText).mkString("|")
+    )
+  }
+
+  protected def writeSkimsForTimePeriods(origins: Seq[GeoUnit], destinations: Seq[GeoUnit], filePath: String): Unit = {
+    //  from activity sim documentation:
+    //    EA - early AM, 3 am to 6 am
+    //    AM - peak period, 6 am to 10 am,
+    //    MD - midday period, 10 am to 3 pm,
+    //    PM - peak period, 3 pm to 7 pm,
+    //    EV - evening, 7 pm to 3 am the next day
+
+    val EAHours = (3 until 6).toList
+    val AMHours = (6 until 10).toList
+    val MDHours = (10 until 15).toList
+    val PMHours = (15 until 19).toList
+    val EVHours = (0 until 3).toList ++ (19 until 24).toList
+
+    val pathTypes = ActivitySimPathType.allPathTypes
+    var writer: BufferedWriter = null
+    try {
+      writer = org.matsim.core.utils.io.IOUtils.getBufferedWriter(filePath)
```

Comment:
Why don't we use CsvWriter from BEAM here?

Reply:
I suppose this is because existing AbstractSkimmer and ODSkimmer use this way for writing CSV. I also checked there are a few other places where it is used, probably we can create a ticket to migrate them all to use CsvWriter?

Reply:
I agree. Let's do this within another ticket.